File: workspace/2023-01-17/caesar-cipher.haskell/caesar-cipher.hs
import Data.Char
import Text.Printf

main = do
    let n = 1
    let s = "Hello, World!"
    let encoded = encode n s
    let decoded = decode n encoded
    printf "n = %d\n" n
    printf "s = %s\n" s
    printf "encoded = %s\n" encoded
    printf "decoded = %s\n" decoded

encode :: Int -> String -> String
encode n s = map (rotate n) s

decode :: Int -> String -> String
decode n = encode (26 - n)

rotate :: Int -> Char -> Char
rotate n c | isLower c = rotateLower n c
           | isUpper c = rotateUpper n c
           | otherwise = rotateNonAlphabet n c
    where
        rotateLower = rotateChar $ ord 'a'
        rotateUpper = rotateChar $ ord 'A'
        rotateNonAlphabet n c = c
        rotateChar offset n c = chr $ (ord c + n - offset) `mod` 26 + offset

File: workspace/2023/2023-02-19/caesar-cipher.ruby/caesar_cipher.rb

def encode(text, n)
    chars = text.chars
    encoded_chars = chars.map{|char| rotate(char, n)}
    encoded_chars.join
end

def decode(encoded_text, n)
    encode(encoded_text, 26 - n)
end

def rotate(char, n)
    case
        when lower?(char)
            rotate_alphabet(char, n, 'a'.ord)
        when upper?(char)
            rotate_alphabet(char, n, 'A'.ord)
        else
            char
    end
end

def lower?(char)
    ('a'..'z').include?(char)
end

def upper?(char)
    ('A'..'Z').include?(char)
end

def rotate_alphabet(char, n, offset)
    ((char.ord + n - offset) % 26 + offset).chr
end

